### PR TITLE
expression, executor, plan: rewrite builtin function trim.

### PR DESCRIPTION
**Pull request by SteveZhangBit**
Bug exists in the current version!

The first is that for SQL `TRIM(LEADING FROM 'bar')`, the parser will convert it into function call `trim('bar', nil, ast.TrimLeading)`.
But for SQL  `TRIM(LEADING NULL FROM 'bar')`, MYSQL will get NULL, but in our system, it will treat it as `TRIM(LEADING FROM 'bar')` and return 'bar'.

The second problem is that the type of keyword `LEADING` is ast.TrimDirectionType. In the new architecture, we have to designate the `argType` in `newBaseBuiltinFunctionWithTp`,  and it should be int for ast.TrimDirectionType.

But when building the datum, its type will be classified to other, which will be string in the end. So it will add a cast function over the expression, and will get an error `cannot convert 2(type ast.TrimDirectionType) to string`.

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -1118,75 +1118,145 @@ func (b *builtinUnHexSig) evalString(row []types.Datum) (string, bool, error) {
...
+		argType := args[0].GetType()
+		bf.tp.Flen = argType.Flen
+		if mysql.HasBinaryFlag(argType.Flag) {
+			types.SetBinChsClnFlag(bf.tp)
+		}
+		sig := &builtinTrim3ArgsSig{baseStringBuiltinFunc{bf}}
+		return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+
+	default:
+		return nil, errors.Trace(c.verifyArgs(args))
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: move the check of c.verifyArg() to the beginning of this function.

**Comment by XuHuaiyu** [comment]
LGTM 
PTAL @zz-jason

**Comment by XuHuaiyu** [comment]
please resolve the conflicts.

**Comment by shenli** [comment]
LGTM

**Comment by zz-jason** [comment]
@SteveZhangBit plz resolve conflicts

**Pull request merged**